File: maybefunctors.hs
module Main where

aList ::  [Integer]
aList = [1, 2, 3]

addOne ::  Num a => a -> a
addOne a = a + 1

addTwo ::  Num a => a -> a
addTwo a = a + 2

composedFn ::  Integer -> Integer
composedFn = addOne . addTwo
res ::  [Integer]
res = map composedFn aList

main ::  IO ()
main = do
    print res
    let res2 = fmap composedFn Nothing
    print res2
    let res3 = fmap composedFn (Just 1)
    print res3

File: maybefunctors.js
'use strict';

var aList = [1, 2, 3];

function compose(f, g) {
  return function (x) {
    return f(g(x));
  };
}

function addOne(value) {
  return value + 1;
}

function addTwo(value) {
  return value + 2;
}

console.log(aList.map(compose(addOne, addTwo)));
console.log(aList.map(addTwo).map(addOne));

function mayBe(value, fn) {
  return value === null || value === undefined ? value : fn(value);
}

console.log(mayBe(undefined, compose(addOne, addTwo)));     // returns expected result undefined
console.log(mayBe(mayBe(undefined, addTwo), addOne));       // returns expected result undefined
console.log(mayBe(1, compose(addOne, addTwo)));             // returns expected result 4 
console.log(mayBe(mayBe(1, addTwo), addOne));               // returns expected result 4
